include/picongpu/particles/atomicPhysics/stage/CheckForFieldEnergyOverSubscription.hpp

```diff
@@ -0,0 +1,103 @@
+ */
+
+/** @file check for overSubscription of cells and calculate rejectionProbability for each*/
+
+#pragma once
+
+// need picongpu::atomicPhysics::ElectronHistogram from atomicPhysics.param
+#include "picongpu/defines.hpp"
+#include "picongpu/fields/FieldE.hpp"
+#include "picongpu/particles/atomicPhysics/electronDistribution/LocalHistogramField.hpp"
+#include "picongpu/particles/atomicPhysics/kernel/CheckForFieldEnergyOverSubscription.kernel"
+#include "picongpu/particles/atomicPhysics/localHelperFields/FieldEnergyUseCacheField.hpp"
+#include "picongpu/particles/atomicPhysics/localHelperFields/RejectionProbabilityCacheField_Cell.hpp"
+#include "picongpu/particles/atomicPhysics/localHelperFields/SharedResourcesOverSubscribedField.hpp"
+#include "picongpu/particles/atomicPhysics/localHelperFields/TimeRemainingField.hpp"
+
+#include <pmacc/Environment.hpp>
+#include <pmacc/mappings/kernel/AreaMapping.hpp>
+#include <pmacc/type/Area.hpp>
+
+#include <cstdint>
+
+namespace picongpu::particles::atomicPhysics::stage
+{
+    /** CheckForAndRejectOversubscription atomic physics sub-stage
+     *
+     * check each histogram bin for deltaWeight > weight0, if yes mark bin as over subscribed
+     *
+     * @tparam T_numberAtomicPhysicsIonSpecies specialization template parameter used to prevent compilation of all
+     *  atomicPhysics kernels if no atomic physics species is present.
+     */
+    template<uint32_t T_numberAtomicPhysicsIonSpecies>
+    struct CheckForFieldEnergyOverSubscription
+    {
+        //! call of kernel for every superCell
+        HINLINE void operator()(picongpu::MappingDesc const mappingDesc) const
+        {
+            // full local domain, no guards
+            pmacc::AreaMapping<CORE + BORDER, MappingDesc> mapper(mappingDesc);
+            pmacc::DataConnector& dc = pmacc::Environment<>::get().DataConnector();
+
+            auto& timeRemainingField = *dc.get<
+                picongpu::particles::atomicPhysics::localHelperFields::TimeRemainingField<picongpu::MappingDesc>>(
+                "TimeRemainingField");
+
+            auto& sharedResourcesOverSubscribedField
+                = *dc.get<picongpu::particles::atomicPhysics::localHelperFields::SharedResourcesOverSubscribedField<
+                    picongpu::MappingDesc>>("SharedResourcesOverSubscribedField");
+
+            auto& rejectionProbabilityCacheField_Cell
+                = *dc.get<picongpu::particles::atomicPhysics::localHelperFields::RejectionProbabilityCacheField_Cell<
+                    picongpu::MappingDesc>>("RejectionProbabilityCacheField_Cell");
+
+            auto& eField = *dc.get<FieldE>(FieldE::getName());
+
+            using FieldEnergyUseCacheField
+                = picongpu::particles::atomicPhysics::localHelperFields::FieldEnergyUseCacheField<
+                    picongpu::MappingDesc>;
+            auto& fieldEnergyUseCacheField = *dc.get<FieldEnergyUseCacheField>("FieldEnergyUseCacheField");
```

Comment:
shoudl be `dc.get` and later use `fieldEnergyUseCacheField->` to access methods.

Reply:
fixed